engine/framework/master.go

```diff
@@ -592,13 +593,22 @@ func (m *DefaultBaseMaster) markStateInMetadata(
 	ctx context.Context, code frameModel.MasterState,
 ) error {
 	metaClient := metadata.NewMasterMetadataClient(m.id, m.frameMetaClient)
-	masterMeta, err := metaClient.Load(ctx)
-	if err != nil {
-		return errors.Trace(err)
-	}
+	m.masterMeta.State = code
+	return metaClient.Update(ctx, m.masterMeta.UpdateStateValues())
+}
+
+func (m *DefaultBaseMaster) persistMetaError() {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+	defer cancel()
 
-	masterMeta.State = code
-	return metaClient.Update(ctx, masterMeta)
+	if err := m.errCenter.CheckError(); err != nil {
+		metaClient := metadata.NewMasterMetadataClient(m.id, m.frameMetaClient)
+		m.masterMeta.ErrorMsg = err.Error()
```

Comment:
Be carefully and clear the `ErrorMsg` in other place to avoid error setting the job meta.